src/cpp/fastdds/subscriber/SubscriberImpl.cpp

```diff
@@ -473,11 +476,23 @@ void SubscriberImpl::SubscriberReaderListener::on_liveliness_changed(
 {
     (void)status;
 
-    if (subscriber_->listener_ != nullptr)
+    SubscriberListener* listener = subscriber_->listener_;
+    if (!subscriber_->user_subscriber_->get_status_mask().is_active(StatusMask::liveliness_changed()))
+    {
+        auto participant = subscriber_->get_participant();
+        auto part_listener = const_cast<DomainParticipantListener*>(participant->get_listener());
+        listener = static_cast<SubscriberListener*>(part_listener);
+        if (!participant->get_status_mask().is_active(StatusMask::liveliness_changed()))
+        {
+            listener = nullptr;
+        }
+    }
```

Comment:
I'd like the callback handling to be the same in all entities, which is not right now:
- The DataReader checks its user listener and mask, if it cannot process the callback, forwards to the Subscriber's inner listener `SubscriberReaderListener`
- The Subscriber checks its user listener and mask, if it cannot process the callback, the Subscriber itself checks the Participant's listener and mask and calls the callback of the Subscriber of Participant accordingly.

It would be beautiful if all entities forward the callback to the parent entity's inner listener, and each Entity checks its own listener and mask; or the other way round, that al the checks are done in the lowest entity (DataReader in this case).

I see drawbacks for both solutions: the former adds new callbacks to the parent entity's inner listener that may have little to do with that entity's work, the latter makes the lower entities access all the parent hierarchy to the top.

Not saying that this must be addressed on this PR (wouldn't reject it because of this), but I think it is worth a though, mostly to decide how we want to implement the listener logic correction (task 8717)

Reply:
I would have a `DataReaderListener* get_reader_listener_for(StatusMask status)` on all implementation entities. Each one will check if they have a listener set and the received mask active, and bubble up to their parent when it is not the case. With that, we could have the call to the listener on a single place (the bottom inner listener).

But, as you say, that is worth a separate PR.